lib/response-handler.js

```diff
@@ -0,0 +1,168 @@
+"use strict";
+
+module.exports = (error, logger, parseLinkHeader) => {
+    /**
+     * @typedef {Object} artifactResponse
+     * @property {string} statusCode The HTTP status code.
+     * @property {string} message The status message for the response.
+     * @property {Object} body The response body.
+     * @property {string} location The url of the uploaded artifact.
+     */
+
+    /**
+     * Handles the response gotten from making requests to Shelf.
+     *
+     * @param {http~IncomingMessage} response
+     * @return {Promise.<Object>}
+     */
+    function handleArtifactResponse(response) {
```

Comment:
I don't think a generic thing like this is necessary.  It seems like we tried to combine POST and GET. These two things share some functionality, but doesn't share others.  Its ok to have separate functions for this.  Consider this.

```
/**
 * Handles errors. Check if under 200, or greater than
 * 399. Map error. Throw exception. This is in common with
 * ALL endpoints
 */
function handleErrorResponse(response) {

}

/**
 * Give me back the path part of the link but do a full resolve based
 * off of the URL we requested. Why do this? Links CAN be relative.  So
 * if I put to "/hello/hi/whatup" and I get a link header back "../../sup/yo"
 * the link I want back is "/hello/sup/yo". I already do this in my generic
 * data uploader stuff.
 *
 * This can be used for both link headers and location headers
 */
function resolveLink(url, link) {

}
/**
 * Takes a response, gets link headers, parses them, creates a list
 * of strings.
 */
function resolveLinkHeaders(response) {}
```

Basically everything else is specific to an endpoint or not needed.  I don't need to assign a success message.  I don't need to handle errors separately.

I also don't need a function per endpoint in a common place. Its ok to put that handling in the functions or classes that handle them.

Also some of the logic makes assumptions, such as "200" being a successful download.  This IS true when trying to download an artifact, but not if I were creating an artifact or a new artifact endpoint we created in the future.